rdr_service/offline/sync_consent_files.py

```diff
@@ -42,6 +45,117 @@ class ParticipantPairingInfo:
     site_name: str
 
 
+@dataclass
+class PairingHistoryRecord:
+    org_name: str
+    start_date: datetime
+
+
+class ConsentSyncGuesser:
+    _SYNC_OVERLAP_DAYS_DELTA = timedelta(days=10)
+
+    def __init__(self, session: Session, participant_history_dao: ParticipantHistoryDao):
+        self._session = session
+        self._participant_history_dao = participant_history_dao
+
+    def check_consents(self, files: Collection[ConsentFile], session):
+        # Get the latest pairing information for the participants
+        participant_ids = [file.participant_id for file in files]
+        raw_pairing_data = self._participant_history_dao.get_pairing_history(
+            session=self._session,
+            participant_ids=participant_ids
+        )
+        latest_participant_pairings = {}
+        for history_record in raw_pairing_data:
+            participant_id = history_record.participantId
+            possible_latest_record = PairingHistoryRecord(
+                org_name=history_record.externalId,
+                start_date=history_record.lastModified
+            )
+            if participant_id not in latest_participant_pairings:
+                latest_participant_pairings[participant_id] = possible_latest_record
+            else:
+                currently_stored_pairing = latest_participant_pairings[participant_id]
+                if currently_stored_pairing.start_date <= possible_latest_record.start_date \
+                        and currently_stored_pairing.org_name != possible_latest_record.org_name:
+                    latest_participant_pairings[participant_id] = possible_latest_record
+
+        participant_summaries = ParticipantSummaryDao.get_by_ids_with_session(self._session, participant_ids)
+        for summary in participant_summaries:
+            pairing_info = latest_participant_pairings[summary.participantId]
+            latest_participant_pairings[summary.participantId] = (pairing_info, summary)
+
+        count = 0
+        for file in files:
+            pairing_info, summary = latest_participant_pairings[file.participant_id]
+            sync_date = self.get_sync_date(file=file, summary=summary, latest_pairing_info=pairing_info)
+            if sync_date is not None:
+                file.sync_time = sync_date
+                file.sync_status = ConsentSyncStatus.SYNC_COMPLETE
+
+            count += 1
```

Comment:
Need clarification to approve:
Is `count` supposed to track how many passes through the loop, or how many file records were actually updated to SYNC_COMPLETE?

Also:   what happens when the count is not a multiple of 500?  Do you need a commit() outside of the `for` loop if there is a remainder value?

Reply:
If there's a remainder value then that will still commit when the session's context closes.